lldb/include/lldb/Host/JSONTransport.h

```diff
@@ -54,112 +50,220 @@ class TransportUnhandledContentsError
-          if (num_bytes == 0) {
-            if (m_buffer.empty())
-              read_cb(loop, llvm::make_error<TransportEOFError>());
-            else
-              read_cb(loop, llvm::make_error<TransportUnhandledContentsError>(
-                                std::string(m_buffer)));
-          }
-        },
-        error);
-    if (error.Fail())
-      return error.takeError();
-    return handle;
-  }
+  /// Public for testing purposes, otherwise this should be an implementation
+  /// detail.
+  static constexpr size_t kReadBufferSize = 1024;
 
 protected:
-  template <typename... Ts> inline auto Logv(const char *Fmt, Ts &&...Vals) {
-    Log(llvm::formatv(Fmt, std::forward<Ts>(Vals)...).str());
+  virtual llvm::Expected<std::vector<std::string>> Parse() = 0;
+  virtual std::string Encode(const llvm::json::Value &message) = 0;
+  void Write(const llvm::json::Value &message) {
+    this->Logv("<-- {0}", message);
+    std::string output = Encode(message);
+    size_t bytes_written = output.size();
+    Status status = m_out->Write(output.data(), bytes_written);
+    if (status.Fail()) {
+      this->Logv("writing failed: s{0}", status.AsCString());
+    }
   }
-  virtual void Log(llvm::StringRef message);
 
-  virtual llvm::Error WriteImpl(const std::string &message) = 0;
-  virtual llvm::Expected<std::vector<std::string>> Parse() = 0;
+  llvm::SmallString<kReadBufferSize> m_buffer;
 
-  static constexpr size_t kReadBufferSize = 1024;
+private:
+  void OnRead(llvm::Error *err, MainLoopBase &loop,
+              typename Transport<Req, Resp, Evt>::MessageHandler &handler) {
+    llvm::ErrorAsOutParameter ErrAsOutParam(err);
+    char buf[kReadBufferSize];
+    size_t num_bytes = sizeof(buf);
+    if (Status status = m_in->Read(buf, num_bytes); status.Fail()) {
+      *err = status.takeError();
+      loop.RequestTermination();
+      return;
+    }
+
+    if (num_bytes)
+      m_buffer.append(llvm::StringRef(buf, num_bytes));
+
+    // If the buffer has contents, try parsing any pending messages.
+    if (!m_buffer.empty()) {
+      llvm::Expected<std::vector<std::string>> raw_messages = Parse();
+      if (llvm::Error error = raw_messages.takeError()) {
+        *err = std::move(error);
+        loop.RequestTermination();
+        return;
+      }
+
+      for (const auto &raw_message : *raw_messages) {
+        auto message =
+            llvm::json::parse<typename Transport<Req, Resp, Evt>::Message>(
+                raw_message);
+        if (!message) {
+          *err = message.takeError();
+          loop.RequestTermination();
+          return;
+        }
+
+        if (Evt *evt = std::get_if<Evt>(&*message)) {
+          handler.OnEvent(*evt);
+        } else if (Req *req = std::get_if<Req>(&*message)) {
+          handler.OnRequest(*req);
+        } else if (Resp *resp = std::get_if<Resp>(&*message)) {
+          handler.OnResponse(*resp);
+        } else {
+          llvm_unreachable("unknown message type");
+        }
```

Comment:
Either no braces or, alternatively you could do something like

```
if (Evt *evt = std::get_if<Evt>(&*message)) {
  handler.OnEvent(*evt);
  continue;
}

if (Req *req = std::get_if<Req>(&*message)) {
  handler.OnRequest(*req);
  continue;
}
```

Reply:
I went with the ifs + `continue`, I think that looks a bit nicer.